x-pack/plugin/autoscaling/src/main/java/org/elasticsearch/xpack/autoscaling/storage/ReactiveStorageDeciderService.java

```diff
@@ -357,24 +417,26 @@ private boolean allocatedToTier(ShardRouting s, RoutingAllocation allocation) {
          * Check that disk decider is only decider for a node preventing allocation of the shard.
          * @return true if and only if a node exists in the tier where only disk decider prevents allocation
          */
-        private boolean cannotAllocateDueToStorage(ShardRouting shard, RoutingAllocation allocation) {
+        private Optional<ShardNodeDecisions> cannotAllocateDueToStorage(ShardRouting shard, RoutingAllocation allocation) {
             if (nodeIds.isEmpty() && needsThisTier(shard, allocation)) {
-                return true;
+                return Optional.of(new ShardNodeDecisions(shard, List.of()));
             }
             assert allocation.debugDecision() == false;
             // enable debug decisions to see all decisions and preserve the allocation decision label
             allocation.debugDecision(true);
             try {
-                boolean diskOnly = nodesInTier(allocation.routingNodes()).map(
-                    node -> allocationDeciders.canAllocate(shard, node, allocation)
-                ).anyMatch(ReactiveStorageDeciderService::isDiskOnlyNoDecision);
-                if (diskOnly && shard.unassigned() && shard.recoverySource().getType() == RecoverySource.Type.LOCAL_SHARDS) {
+                List<NodeDecision> diskOnly = nodesInTier(allocation.routingNodes()).map(
+                    node -> new NodeDecision(node.node(), allocationDeciders.canAllocate(shard, node, allocation))
+                ).filter(nar -> ReactiveStorageDeciderService.isDiskOnlyNoDecision(nar.decision())).toList();
+                if (diskOnly.size() > 0 && shard.unassigned() && shard.recoverySource().getType() == RecoverySource.Type.LOCAL_SHARDS) {
                     // For resize shards only allow autoscaling if there is no other node where the shard could fit had it not been
                     // a resize shard. Notice that we already removed any initial_recovery filters.
-                    diskOnly = nodesInTier(allocation.routingNodes()).map(node -> allocationDeciders.canAllocate(shard, node, allocation))
-                        .anyMatch(ReactiveStorageDeciderService::isResizeOnlyNoDecision) == false;
+                    List<NodeDecision> resizeOnly = nodesInTier(allocation.routingNodes()).map(
+                        node -> new NodeDecision(node.node(), allocationDeciders.canAllocate(shard, node, allocation))
+                    ).filter(nar -> ReactiveStorageDeciderService.isResizeOnlyNoDecision(nar.decision())).toList();
+                    return resizeOnly.isEmpty() ? Optional.of(new ShardNodeDecisions(shard, resizeOnly)) : Optional.empty();
```

Comment:
I think this is incorrect, we should return the diskOnly decision here as well, otherwise we won't know why we cannot allocate the shard, right?

Reply:
You're totally right, I've changed the check, so we have one return statement with the disk only node decisions